keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,1087 @@
+    - matchExpressions:
+      - key: failure-domain.beta.kubernetes.io/region
+        operator: In
+        values:
+        - us-east-1
+
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-64103396-0d32-11ea-945c-e3ede5f0f3ae
+spec:
+  driverName: pd.csi.storage.gke.io
+status:
+  classes:
+  - maximumVolumeSize: 256G
+    storageClassName: some-storage-class
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: failure-domain.beta.kubernetes.io/region
+        operator: In
+        values:
+        - us-west-1
+```
+
+#### CSIDriver.spec.storageCapacity
+
+A new field `storageCapacity` of type `boolean` with default `false`
+in
+[CSIDriver.spec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#csidriverspec-v1beta1-storage-k8s-io)
+indicates whether a driver deployment will create `CSIStoragePool`
+objects with capacity information and wants the Kubernetes scheduler
+to rely on that information when making scheduling decisions that
+involve volumes that need to be created by the driver.
+
+If not set, the scheduler makes such decisions without considering
+whether the driver really can create the volumes (the current situation).
+
+
+### Updating capacity information with external-provisioner
+
+Most (if not all) CSI drivers already get deployed on Kubernetes
+together with the external-provisioner which then handles volume
+provisioning via PVC. However, drivers which only handle inline
+ephemeral volumes currently don’t need that and thus get deployed
+without it.
+
+For the sake of avoiding yet another sidecar, the proposal is to
+extend the external-provisioner such that it can also be deployed
+alongside a CSI driver on each node and then just handles
+`CSIStoragePool` creation and updating. This leads to different modes
+of operations.
+
+Because the external-provisioner is part of the deployment of the CSI
+driver, that deployment can configure the behavior of the
+external-provisioner via command line parameters. There is no need to
+introduce heuristics or other, more complex ways of changing the
+behavior (like extending the `CSIDriver` API).
+
+#### Available capacity vs. maximum volume size
+
+The CSI spec up to and including the current version 1.2 just
+specifies that ["the available
+capacity"](https://github.com/container-storage-interface/spec/blob/314ac542302938640c59b6fb501c635f27015326/lib/go/csi/csi.pb.go#L2548-L2554)
+is to be returned by the driver. It is left open whether that means
+that a volume of that size can be created.
+
+This KEP introduces the more precise definition of a "maximum volume
+size". Unless told otherwise, the external-provisioner assumes that
+the `GetCapacityResponse` is the sum of all currently available
+storage and then sets `CSIStorageByClass.AvailableCapacity` instead of
+`CSIStorageByClass.MaximumVolumeSize`. A CSI driver deployment can use
+`--capacity-is-maximum-volume-size=true` to change that, in which
```

Comment:
I don't see value in adding this option at the moment because we're not interpreting availablecapacity and maxvolumesize differently in the scheduler. And it's one more thing where we're working around the csi spec. I'd like to avoid creating more peculiarities and differences between K8s and CSI because it has never worked out well in the past.